### PR TITLE
replicaset: add `expel` command

### PR DESCRIPTION
**Pull request by askalt**
@TarantoolBot document
Title: new commands: `tt replicaset expel`, `tt cluster replicaset expel`

The patch introduces a new command for the `tt replicaset` module:

* `tt replicaset expel APP_NAME:INSTANCE_NAME` - to expel an instance from a cluster.

The command automatically determines a replicaset orchestrator. Currently, it supports:

* cartridge - the Cartridge orchestrator for Tarantool <= 2.
* config - the centralized config orchestrator for Tarantool >= 3.

Usage:

tt replicaset expel [--cartridge|--config|--custom] <APP_NAME:INSTANCE_NAME>

The example of execution:

```
$ tt replicaset expel myapp:instance-003
   • Discovery application...

Orchestrator:      centralized config
Replicasets state: bootstrapped

• replicaset-001
  Failover: off
  Master:   single
    • instance-001 unix/:./instance-001.iproto rw
    • instance-002 unix/:./instance-002.iproto read
    • instance-003 unix/:./instance-003.iproto read
• replicaset-002
  Failover: off
  Master:   multi
    • instance-004 unix/:./instance-004.iproto rw
    • instance-005 unix/:./instance-005.iproto rw

   • Expel instance: instance-003
   • Done.
$ tt replicaset status myapp
Orchestrator:      centralized config
Replicasets state: bootstrapped

• replicaset-001
  Failover: off
  Master:   single
    • instance-001 unix/:./instance-001.iproto rw
    • instance-002 unix/:./instance-002.iproto read
• replicaset-002
  Failover: off
  Master:   multi
    • instance-004 unix/:./instance-004.iproto rw
    • instance-005 unix/:./instance-005.iproto rw
```

Also, new command for `cluster` module:
* `tt cluster replicaset expel [-f] [flags] <URI> <INSTANCE_NAME>` expels an instance from the 3.0 replicaset by patching the config in storage (etcd or tarantool config storage). 

The example of execution:
```sh
$ tt cluster show http://localhost:2379/foo

credentials:
  users:
    guest:
      roles:
        - super
groups:
  group-001:
    replicasets:
      replicaset-001:
        instances:
          instance-001:
            iproto:
              listen:
                - uri: localhost:3301
          instance-002:
            iproto:
              listen:
                - uri: localhost:3302

$ tt cluster rs expel http://localhost:2379/foo instance-002
   • Patch the config by the key: "/foo/config/all"
   • Done.   

$ tt cluster show http://localhost:2379/foo
credentials:
  users:
    guest:
      roles:
        - super
groups:
  group-001:
    replicasets:
      replicaset-001:
        instances:
          instance-001:
            iproto:
              listen:
                - uri: localhost:3301
          instance-002:
            iproto:
              listen: {}
```

Closes https://github.com/tarantool/tt/issues/318

### CODE REVIEW

File: cli/replicaset/cartridge.go [hunk 1]
```diff
@@ -47,6 +50,7 @@ type cartridgeTopology struct {
 
 // CartridgeInstance is an instance with the Cartridge orchestrator.
 type CartridgeInstance struct {
+	cachedDiscoverer
 	evaler connector.Evaler
 }
```
[the remark this answers is not in the document]
Comment (oleg-jukovec) [reply cli/replicaset/cartridge.go:55]: I propose to do it in that way: https://go.dev/play/p/vEQPpj1HEsL
It is a little tricky too, but a bit more clear in my taste.

File: cli/replicaset/cartridge.go [hunk 2]
```diff
@@ -59,9 +63,11 @@ type cartridgeInstance struct {
 
 // NewCartridgeInstance creates a new CartridgeInstance object for the evaler.
 func NewCartridgeInstance(evaler connector.Evaler) *CartridgeInstance {
-	return &CartridgeInstance{
+	inst := &CartridgeInstance{
 		evaler: evaler,
 	}
+	inst.cachedDiscoverer = newCachedDiscoverer(inst)
+	return inst
```
Comment (oleg-jukovec) [inline cli/replicaset/cartridge.go:70]: I'm about that code.

File: cli/replicaset/cartridge.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (askalt) [reply cli/replicaset/cartridge.go:55]: Applied, thank you!

File: test/integration/replicaset/test_replicaset_expel.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
+
+
```
[the remark this answers is not in the document]
Comment (askalt) [reply test/integration/replicaset/test_replicaset_expel.py]: Removed.

File: cli/cmd/replicaset.go [hunk 4]
```diff
@@ -324,6 +353,32 @@ func internalReplicasetStatusModule(cmdCtx *cmdcontext.CmdCtx, args []string) er
...
+func internalReplicasetExpelModule(cmdCtx *cmdcontext.CmdCtx, args []string) error {
+	if _, _, found := strings.Cut(args[0], string(running.InstanceDelimiter)); !found {
+		return fmt.Errorf("the command expects argument application_name:instance_name")
+	}
+	var ctx replicasetCtx
+	if err := replicasetFillCtx(cmdCtx, &ctx, args, true); err != nil {
+		return err
+	}
+	collectors, publishers, err := createDataCollectorsAndDataPublishers(
+		cmdCtx.Integrity, "")
```
[the remark this answers is not in the document]
Comment (askalt) [reply cli/cmd/replicaset.go]: It's my mistake, the flag for key was added.

File: CHANGELOG.md [hunk 5]
```diff
@@ -13,6 +13,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
  cluster config (3.0) or cartridge orchestrator.
 - `tt replicaset demote`: command to demote an instance in the tarantool replicaset with
  cluster config (3.0) orchestrator.
+- `tt replicaset expel`:  command to expel an instance from the tarantool replicaset with
+ cluster config (3.0) or cartridge orchestrator.
```
[the remark this answers is not in the document]
Comment (askalt) [reply CHANGELOG.md:17]: Yes, added now.

**Review by oleg-jukovec — APPROVED** [review]

File: cli/replicaset/discovery.go [hunk 6]
```diff
@@ -1,69 +1,39 @@
...
-	"github.com/tarantool/tt/cli/connector"
-	"github.com/tarantool/tt/cli/running"
-)
-
 // Discoverer is an interface for discovering information about
 // replicasets.
 type Discoverer interface {
 	// Discovery returns replicasets information or an error.
-	Discovery() (Replicasets, error)
+	Discovery(force bool) (Replicasets, error)
```
Comment (psergee) [inline cli/replicaset/discovery.go:7]: Boolean parameters are harder to maintain. If I see `discoverer.Discovery(true)` somewhere, I have no idea what the `true` does. Let's use an enum or split the function into `Discovery` and `ForcedDiscovery`, whatever that means.

**Pull request merged**